src/wasm/wasm-binary.cpp

```diff
@@ -6920,13 +6920,10 @@ bool WasmBinaryBuilder::maybeVisitRefCast(Expression*& out, uint32_t code) {
     } else {
       nullability = code == BinaryConsts::RefCast ? NonNullable : Nullable;
     }
-    // Only accept instructions emulating the legacy behavior for now.
-    if (ref->type.isRef()) {
-      if (nullability == NonNullable && ref->type.isNullable()) {
-        throwError("ref.cast on nullable input not yet supported");
-      } else if (nullability == Nullable && ref->type.isNonNullable()) {
-        throwError("ref.cast null on non-nullable input not yet supported");
-      }
+    // Implicitly convert nullable casts of non-null references to non-nullable
+    // casts to avoid losing type information.
+    if (ref->type.isRef() && ref->type.isNonNullable()) {
+      nullability = NonNullable;
```

Comment:
Do we need this? It could be left for optimizations.

Reply:
Yes, because otherwise our validation check for this will fail. We could remove the validation rule as well, but I think it's helpful to simply disallow the suboptimal configuration.

Reply:
I see. Makes sense, but it will also need to be in the C API and every location that creates such an instruction. I don't feel strongly though. If you do want to go this route, please document it in the readme among our list of IR differences from wasm.

Reply:
Actually, since finalization already implements this automatic transition away from the undesirable configuration, we probably don't need to implement it anywhere else, even in the parsers. Does that sound right?

Reply:
Good point, yeah, finalization is the proper single place to do this canonicalization. I'd fuzz it carefully though to make sure we're not missing something.

Reply:
Fuzzed 58k iterations with auto-initial-contents over night, so I think this should be ready to go.